NDTensors/src/fill.jl

```diff
@@ -0,0 +1,54 @@
+# This is a file to write generic fills for NDTensors.
+#  This includes random fills, zeros, ...
+
+function generic_randn(
+  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
+) where {DataT<:AbstractArray,ElT}
+  @assert ElT == eltype(DataT)
+  data = generic_randn(DataT, dim)
+  StoreT = set_datatype(StoreT, typeof(data))
+  return StoreT(data)
+end
+
+function generic_randn(StoreT::Type{<:Dense{ElT}}, dim::Integer=0) where {ElT}
+  return generic_randn(default_storagetype(ElT), dim)
+end
+
+function generic_randn(StoreT::Type{<:Dense}, dim::Integer=0)
+  return generic_randn(default_storagetype(), dim)
+end
+
+function generic_randn(DataT::Type{<:AbstractArray}, dim::Integer=0)
+  DataT = NDTensors.set_eltype_if_unspecified(DataT)
+  data = similar(DataT, dim)
+  ElT = eltype(DataT)
+  for i in 1:length(data)
+    data[i] = randn(ElT)
+  end
+  return data
+end
+
+##TODO replace randn in ITensors with generic_randn
+## and replace zeros with generic_zeros
+function generic_zeros(
+  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
+) where {DataT<:AbstractArray,ElT}
+  @assert ElT == eltype(DataT)
```

Comment:
Do we need this check here? Can't this be handled by the inner constructor for `Dense`?

Reply:
If you remove the assert, this is what happens
```
julia> NDTensors.generic_zeros(NDTensors.Dense{Float32, Vector{Float64}}, 20)
20-element NDTensors.Dense{Float64, Vector{Float64}}:
```
My argument is that, this isn't correct because theres no way to know if `ElT` or `eltype(DataT)` is the correct type.

Reply:
Yes I agree we don't want that behavior, I was just hoping that we could catch these kinds of mistakes farther down in an inner constructor so that we wouldn't have to spread checks like that around the codebase. But I see that because of the line:
```julia
StoreT = set_datatype(StoreT, typeof(data))
```
`StoreT` would start with contradictory element types but then `set_datatype` would fix it to the eltype of `data`/`DataT`. Maybe it should be `set_datatype` that checks for inconsistent element types?

It doesn't have to be dealt with in this PR, but we can think about good practices for catching inconsistent type parameters in future PRs. Something we could think about is having a catch-all generic function `check_consistent_type_parameters` that defines checks like:
```julia
function check_consistent_type_parameters(storagetype::Type{<:TensorStorage})
  @assert eltype(storagetype) === eltype(datatype(storagetype))
end
```